RecoLocalCalo/HcalRecProducers/src/HBHEPhase1Reconstructor.cc

```diff
@@ -121,6 +127,48 @@ namespace {
     {
         return std::pair<bool,bool>(df.linkError(), df.capidError());
     }
+
+    std::unique_ptr<HBHEStatusBitSetter> parse_HBHEStatusBitSetter(
+        const edm::ParameterSet psdigi)
+    {
+        return std::make_unique<HBHEStatusBitSetter>(
+            psdigi.getParameter<double>("nominalPedestal"),
+            psdigi.getParameter<double>("hitEnergyMinimum"),
+            psdigi.getParameter<int>("hitMultiplicityThreshold"),
+            psdigi.getParameter<std::vector<edm::ParameterSet> >("pulseShapeParameterSets"));
+    }
+
+    std::unique_ptr<HBHEPulseShapeFlagSetter> parse_HBHEPulseShapeFlagSetter(
+        const edm::ParameterSet &psPulseShape)
+    {
+        return std::make_unique<HBHEPulseShapeFlagSetter>(
+            psPulseShape.getParameter<double>("MinimumChargeThreshold"),
+            psPulseShape.getParameter<double>("TS4TS5ChargeThreshold"),
```

Comment:
I suggest reading the description of "std::make_unique" in a good C++ book.
This is not a "normal" function call. It is, instead, a constructor replacement.

On 09/08/2016 08:10 AM, Slava Krutelyov wrote:
>
> In RecoLocalCalo/HcalRecProducers/src/HBHEPhase1Reconstructor.cc <https://github.com/cms-sw/cmssw/pull/15779#discussion_r78001850>:
>
> > +    std::unique_ptr<HBHEStatusBitSetter> parse_HBHEStatusBitSetter(
> > +        const edm::ParameterSet psdigi)
> > +    {
> > +        return std::make_unique<HBHEStatusBitSetter>(
> > +            psdigi.getParameter<double>("nominalPedestal"),
> > +            psdigi.getParameter<double>("hitEnergyMinimum"),
> > +            psdigi.getParameter<int>("hitMultiplicityThreshold"),
> > +            psdigi.getParameter<std::vector<edm::ParameterSet> >("pulseShapeParameterSets"));
> > +    }
> > +
> > +    std::unique_ptr<HBHEPulseShapeFlagSetter> parse_HBHEPulseShapeFlagSetter(
> > +        const edm::ParameterSet &psPulseShape)
> > +    {
> > +        return std::make_unique<HBHEPulseShapeFlagSetter>(
> > +            psPulseShape.getParameter<double>("MinimumChargeThreshold"),
> > +            psPulseShape.getParameter<double>("TS4TS5ChargeThreshold"),
>
> as another refactoring item for the future,
> may I ask to replace function calls with N>~5 arguments by calls using a struct. So that each argument is easily traceable when reading the code or debugging, or when the list needs an update to 
> remove/add/rename an argument in the future developments.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/cms-sw/cmssw/pull/15779/files/d201d93f931d7a63c81f8689979d4ffd6cdedbeb#r78001850>, or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AFS4-aLW0x-PgGKatyno47czyEkkuW7zks5qoAlVgaJpZM4J36UI>.
>